bloom-housing/bloom
# refactor: preferences & programs data model merged

# Pull Request Template

## Issue Overview

This PR addresses #2752


## Description

This PR updates the preferences and programs data model to be able to support the [Figma designs](https://www.figma.com/file/Y8bFd0aSz7eYJ2TCNid6s3/bloom-partners?node-id=5725%3A182890) for the new preferences and programs management. It also merges the preference and program data models since they were nearly identical, and since the new designs for managing them are nearly identical.

Overview of the refactor:
The preference and program entities have been merged into a new entity that encompasses both of them called `Multiselect Question`
MultiselectQuestion
* Previously, preferences had their titles and option titles stored as just keys that map into our translation files. To support the new design, these need to be just English strings.
* The multiselect question has a new field called `applicationSection` to distinguish what section of an application it belongs in, right now we just have `preferences` and `programs`.

MultiselectOption
* A preference set had its options stored in a `formMetadata` field - this is now a new object called `options`, and the root setting of `hideFromListing` has been moved up to the question level. We no longer support custom select text, and don't need `hideGenericDecline` due to our new design around adding opt out text.
* An individual option now supports links.
* The data types for what sorts of extra data an option can collect has been reduced to just an address.
* The exclusive flag if true indicates that when that option is selected, it can only be selected exclusively, so every other option must automatically uncheck. Our decline options behave this way.

## How Can This Be Tested/Reviewed?

There are four main areas where we use preferences and programs that need to be tested:
(1) Public Listing View
* Should display all preferences and their links/description unless the `hideFromListing` tag is true

(2) Public Application
* Should now have a new section in the application labelled `Programs` that is distinct from `Preferences` if the application does have programs.
* If every option in a multiselect question is exclusive, render as a radio question
* The application summary page before submission should render the answers as expected

(3) Partners Application
* Data for preferences and programs when submitted should be reflected accurately
* The paper application form on partners should submit accurately
* The CSV export should display the data for preferences and programs accurately

(4) Partners Listing Form
* Jurisdictional preferences and programs should be appropriately fetched when selected them on a listing
* They should both save accurately

## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
